In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics.pairwise import cosine_similarity, euclidean_distances, manhattan_distances

In [2]:
df = pd.read_csv('1980.csv')

In [3]:
df['genres'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'bluegrass', 'blues', 'brazil', 'british', 'cantopop',
       'chicago-house', 'children', 'chill', 'classical', 'comedy',
       'country', 'dance', 'dancehall', 'deep-house', 'disco', 'dub',
       'dubstep', 'edm', 'electro', 'electronic', 'folk', 'forro',
       'french', 'funk', 'garage', 'german', 'gospel', 'goth', 'groove',
       'grunge', 'guitar', 'happy', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'indian', 'indie',
       'indie-pop', 'industrial', 'j-idol', 'j-pop', 'j-rock', 'jazz',
       'k-pop', 'latin', 'latino', 'malay', 'mandopop', 'metal',
       'minimal-techno', 'mpb', 'new-age', 'opera', 'pagode', 'party',
       'piano', 'pop', 'pop-film', 'power-pop', 'progressive-house',
       'psych-rock', 'punk', 'punk-rock', 'r-n-b', 'reggae', 'reggaeton',
       'rock', 'rock-n-roll', 'rockabilly', 'salsa', 'samba', 'sertanejo',
       '

Check for duplicate values

In [4]:
df.shape

(293997, 24)

In [5]:
df.drop_duplicates(inplace=True)

In [6]:
df.shape

(76414, 24)

There were a lot of duplicate values..

Check for null values

In [7]:
df.isna().any()

name                      False
artists                   False
album                     False
release_date              False
release_date_precision    False
chart_power                True
spotify_id                False
uri                       False
popularity                False
danceability              False
energy                    False
key                       False
loudness                  False
mode                      False
speechiness               False
acousticness              False
instrumentalness          False
liveness                  False
valence                   False
tempo                     False
duration_ms               False
time_signature            False
isrc                       True
genres                    False
dtype: bool

There exist Null values for the Chart Power and for the isrc. The null values for the Chart Power are valid, because not every song was in the Charts. Lets have a look at the null values for isrc:

In [8]:
df[df['isrc'].isna()]

name       artists           album release_date  \
171526  Amor de Verano  Pintura Roja  Amor de Verano         1980   

       release_date_precision  chart_power              spotify_id  \
171526                   year          NaN  2ndwh9yRCFK3CA5zhaJWXY   

                                         uri  popularity  danceability  ...  \
171526  spotify:track:2ndwh9yRCFK3CA5zhaJWXY          32         0.515  ...   

        speechiness  acousticness  instrumentalness  liveness  valence  \
171526       0.0352       0.00685             0.706    0.0737    0.909   

          tempo  duration_ms  time_signature  isrc  genres  
171526  105.044       204202               4   NaN     pop  

[1 rows x 24 columns]

Select all categorical features and group by the spotify id. All values of the features are stored as lists.

In [9]:
df_categorical = df[['spotify_id', 'genres', 'name', 'artists', 'album', 'release_date', 'release_date_precision', 'uri', 'isrc']]

In [10]:
display(df_categorical.head())
df_categorical.shape

spotify_id    genres                                     name  \
0  5xgPirJuHEYxM3pMFPYtXG  acoustic  My Baby Has Eyes Like An Eagle - Take 2   
1  6vonkFQZRt9WHHfO6iU1AL  acoustic                          Round And Round   
2  7pFvU6cr3CURlNc8S9SxAp  acoustic                      Ain't Doin' No Good   
3  6CR1NaaMBPQUed2i5QRW4l  acoustic                     Gonna Call the Angel   
4  1td7cg3olnhudXRKSASjRZ  acoustic       Gonna Call the Angel - Master Take   

                     artists                              album release_date  \
0  Mississippi Fred McDowell                   Levee Camp Blues   1980-01-01   
1                J.B. Lenoir                Down in Mississippi         1980   
2              Johnny Shines  Johnny Shines and Robert Lockwood   1980-01-14   
3              Johnny Shines  Johnny Shines and Robert Lockwood   1980-01-14   
4              Johnny Shines  Johnny Shines and Robert Lockwood   1980-01-14   

  release_date_precision                                   uri          isrc  
0                    day  spotify:track:5xgPirJuHEYxM3pMFPYtXG  US2AH0401019  
1                   year  spotify:track:6vonkFQZRt9WHHfO6iU1AL  DER387900019  
2                    day  spotify:track:7pFvU6cr3CURlNc8S9SxAp  US2Y30629315  
3                    day  spotify:track:6CR1NaaMBPQUed2i5QRW4l  US2Y30629323  
4                    day  spotify:track:1td7cg3olnhudXRKSASjRZ  US2Y30629321

(76414, 9)

In [11]:
df_categorical = df_categorical.groupby('spotify_id').agg(lambda x: list(x)).reset_index()

In [12]:
display(df_categorical.head())
df_categorical.shape

spotify_id               genres  \
0  000P83HDtOHcNVFZy7Q2Yu              [salsa]   
1  000paUq6xoYcvCz2REJdu1      [pop, pop-film]   
2  0017A9myIXhXYB4gBsPRsA                [pop]   
3  001reycpvPkUjmpxTJv6eV  [classical, german]   
4  002H0J6N0wKjjRtFQhKDzZ     [dance, turkish]   

                                                name  \
0                           [El Licor de Tu Boquita]   
1  [Hun Tey Kharkan Gay Alarum, Hun Tey Kharkan G...   
2                                [Untuk Sebuah Nama]   
3  [Oboe Concerto in F Minor, TWV 51:f1: I. Alleg...   
4         [Arabesk Oyun Havası, Arabesk Oyun Havası]   

                                             artists  \
0                     [El Gran Combo De Puerto Rico]   
1                           [Noor Jehan, Noor Jehan]   
2                                   [Ratih Purwasih]   
3  [Georg Philipp Telemann,Burkhard Glaetzner,Har...   
4     [Esin Engin Orkestrası, Esin Engin Orkestrası]   

                                               album  \
0                                            [Unity]   
1  [Noor Jehan New Punjabi Songs, Noor Jehan New ...   
2                                [Untuk Sebuah Nama]   
3  [Oboe Concertos (Baroque) - Antonio Vivaldi: G...   
4                           [Arabesk'80, Arabesk'80]   

               release_date release_date_precision  \
0                    [1980]                 [year]   
1  [1980-12-01, 1980-12-01]             [day, day]   
2                    [1980]                 [year]   
3              [1980, 1980]           [year, year]   
4              [1980, 1980]           [year, year]   

                                                 uri  \
0             [spotify:track:000P83HDtOHcNVFZy7Q2Yu]   
1  [spotify:track:000paUq6xoYcvCz2REJdu1, spotify...   
2             [spotify:track:0017A9myIXhXYB4gBsPRsA]   
3  [spotify:track:001reycpvPkUjmpxTJv6eV, spotify...   
4  [spotify:track:002H0J6N0wKjjRtFQhKDzZ, spotify...   

                           isrc  
0                [USMRE0801806]  
1  [USA561127553, USA561127553]  
2                [UKFT41826340]  
3  [DES380744015, DES380744015]  
4  [BGA472127386, BGA472127386]

(38456, 9)

Select all numerical features and group them by the spotify id.

In [13]:
df_numerical = df.drop(columns=['genres', 'name', 'artists', 'album', 'release_date', 'release_date_precision', 'uri', 'isrc'])

In [14]:
df_numerical.drop_duplicates(inplace=True)

In [15]:
df_numerical.head()

chart_power              spotify_id  popularity  danceability  energy  key  \
0          NaN  5xgPirJuHEYxM3pMFPYtXG           8         0.600   0.118    3   
1          NaN  6vonkFQZRt9WHHfO6iU1AL           5         0.887   0.344    1   
2          NaN  7pFvU6cr3CURlNc8S9SxAp           1         0.433   0.197    8   
3          NaN  6CR1NaaMBPQUed2i5QRW4l           1         0.442   0.167    2   
4          NaN  1td7cg3olnhudXRKSASjRZ           0         0.468   0.205    2   

   loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
0   -18.452     1       0.0402         0.977          0.072900    0.1100   
1   -11.985     1       0.0481         0.832          0.028000    0.3370   
2   -13.150     1       0.0324         0.993          0.003270    0.1090   
3   -14.996     1       0.0358         0.859          0.001730    0.0876   
4   -13.606     0       0.0491         0.778          0.000237    0.0832   

   valence    tempo  duration_ms  time_signature  
0    0.557   98.429       278787               4  
1    0.620  129.867       171445               4  
2    0.804   85.316       155840               4  
3    0.243   72.212       189427               4  
4    0.374  114.904       209107               3

In [16]:
df_numerical.shape

(39520, 16)

In [17]:
df_numerical = df_numerical.groupby('spotify_id').mean()

In [18]:
df_numerical.shape

(38456, 15)

In [19]:
df_numerical.reset_index(inplace=True)

Merge both DataFrames grouped by the spotify id together.

In [20]:
df_total = df_categorical.merge(df_numerical, how='left', on='spotify_id')

In [21]:
df_total

spotify_id                         genres  \
0      000P83HDtOHcNVFZy7Q2Yu                        [salsa]   
1      000paUq6xoYcvCz2REJdu1                [pop, pop-film]   
2      0017A9myIXhXYB4gBsPRsA                          [pop]   
3      001reycpvPkUjmpxTJv6eV            [classical, german]   
4      002H0J6N0wKjjRtFQhKDzZ               [dance, turkish]   
...                       ...                            ...   
38451  7zsH8mPq4nPBqqqDJpjgJe                          [pop]   
38452  7zv3sIP0Umtz9MswoJSF1T                      [country]   
38453  7zwHfIlTAIPa6hgjzK1MGx                         [folk]   
38454  7zwlOlX3MA7BFkmf0c1y4p                 [guitar, jazz]   
38455  7zzdiBe3usdBtyqHEekCEs  [dance, pop, rock, synth-pop]   

                                                    name  \
0                               [El Licor de Tu Boquita]   
1      [Hun Tey Kharkan Gay Alarum, Hun Tey Kharkan G...   
2                                    [Untuk Sebuah Nama]   
3      [Oboe Concerto in F Minor, TWV 51:f1: I. Alleg...   
4             [Arabesk Oyun Havası, Arabesk Oyun Havası]   
...                                                  ...   
38451                                     [Nyalakan Api]   
38452                       [O Little Town of Bethlehem]   
38453           [Mac's fancy, The Mist Covered Mountain]   
38454  [Bianca's Midnight Lullaby, Bianca's Midnight ...   
38455  [Capital City - Remastered 2002, Capital City ...   

                                                 artists  \
0                         [El Gran Combo De Puerto Rico]   
1                               [Noor Jehan, Noor Jehan]   
2                                       [Ratih Purwasih]   
3      [Georg Philipp Telemann,Burkhard Glaetzner,Har...   
4         [Esin Engin Orkestrası, Esin Engin Orkestrası]   
...                                                  ...   
38451                                      [Deddy Dores]   
38452                                      [Johnny Cash]   
38453                                        [De Dannan]   
38454                         [Al Di Meola, Al Di Meola]   
38455  [Simple Minds, Simple Minds, Simple Minds, Sim...   

                                                   album  \
0                                                [Unity]   
1      [Noor Jehan New Punjabi Songs, Noor Jehan New ...   
2                                    [Untuk Sebuah Nama]   
3      [Oboe Concertos (Baroque) - Antonio Vivaldi: G...   
4                               [Arabesk'80, Arabesk'80]   
...                                                  ...   
38451                                     [Nyalakan Api]   
38452                                [Classic Christmas]   
38453                        [The Mist Covered Mountain]   
38454                 [Splendido Hotel, Splendido Hotel]   
38455  [Empires And Dance, Empires And Dance, Empires...   

                                           release_date  \
0                                                [1980]   
1                              [1980-12-01, 1980-12-01]   
2                                                [1980]   
3                                          [1980, 1980]   
4                                          [1980, 1980]   
...                                                 ...   
38451                                            [1980]   
38452                                      [1980-12-01]   
38453                                      [1980-01-01]   
38454                                      [1980, 1980]   
38455  [1980-09-01, 1980-09-01, 1980-09-01, 1980-09-01]   

      release_date_precision  \
0                     [year]   
1                 [day, day]   
2                     [year]   
3               [year, year]   
4               [year, year]   
...                      ...   
38451                 [year]   
38452                  [day]   
38453                  [day]   
38454           [year, year]   
38455   [day, day, day, da

Create a DataFrame which contains only the relevant features for the recommender system

In [22]:
df_recommender = df.drop(columns=['artists', 'genres','album', 'release_date', 'release_date_precision', 'chart_power', 'uri', 'popularity', 'name', 'isrc'])

Drop duplicate values..

In [23]:
df_recommender.shape

(76414, 14)

In [24]:
df_recommender.drop_duplicates(inplace=True)

In [25]:
df_recommender.shape

(38456, 14)

There were a lot of duplicates..

In [26]:
df_recommender.head()

spotify_id  danceability  energy  key  loudness  mode  \
0  5xgPirJuHEYxM3pMFPYtXG         0.600   0.118    3   -18.452     1   
1  6vonkFQZRt9WHHfO6iU1AL         0.887   0.344    1   -11.985     1   
2  7pFvU6cr3CURlNc8S9SxAp         0.433   0.197    8   -13.150     1   
3  6CR1NaaMBPQUed2i5QRW4l         0.442   0.167    2   -14.996     1   
4  1td7cg3olnhudXRKSASjRZ         0.468   0.205    2   -13.606     0   

   speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
0       0.0402         0.977          0.072900    0.1100    0.557   98.429   
1       0.0481         0.832          0.028000    0.3370    0.620  129.867   
2       0.0324         0.993          0.003270    0.1090    0.804   85.316   
3       0.0358         0.859          0.001730    0.0876    0.243   72.212   
4       0.0491         0.778          0.000237    0.0832    0.374  114.904   

   duration_ms  time_signature  
0       278787               4  
1       171445               4  
2       155840               4  
3       189427               4  
4       209107               3

Since all features except the spotify_id are numerical features and the spotify_id should be an unique identifier, try to group by the id to reduce the DataFrame even more.

In [27]:
df_recommender = df_recommender.groupby(['spotify_id']).mean()

In [28]:
display(df_recommender.head())
df_recommender.shape

danceability  energy  key  loudness  mode  \
spotify_id                                                          
000P83HDtOHcNVFZy7Q2Yu         0.619   0.561  7.0    -8.377   0.0   
000paUq6xoYcvCz2REJdu1         0.492   0.558  2.0    -9.944   1.0   
0017A9myIXhXYB4gBsPRsA         0.220   0.668  0.0    -3.549   1.0   
001reycpvPkUjmpxTJv6eV         0.448   0.224  5.0   -14.759   0.0   
002H0J6N0wKjjRtFQhKDzZ         0.286   0.480  2.0   -12.444   0.0   

                        speechiness  acousticness  instrumentalness  liveness  \
spotify_id                                                                      
000P83HDtOHcNVFZy7Q2Yu       0.0736         0.438          0.000042     0.126   
000paUq6xoYcvCz2REJdu1       0.0868         0.741          0.037700     0.222   
0017A9myIXhXYB4gBsPRsA       0.0289         0.084          0.000004     0.117   
001reycpvPkUjmpxTJv6eV       0.0348         0.831          0.916000     0.129   
002H0J6N0wKjjRtFQhKDzZ       0.0365         0.528          0.323000     0.343   

                        valence    tempo  duration_ms  time_signature  
spotify_id                                                             
000P83HDtOHcNVFZy7Q2Yu    0.696   88.940     242493.0             4.0  
000paUq6xoYcvCz2REJdu1    0.606   96.169     372167.0             4.0  
0017A9myIXhXYB4gBsPRsA    0.360   81.433     249504.0             4.0  
001reycpvPkUjmpxTJv6eV    0.540  112.185     215733.0             4.0  
002H0J6N0wKjjRtFQhKDzZ    0.752  134.274     345943.0             4.0

(38456, 13)

The shape stayed the same, so there were no duplicate ids.

Scale the data, so every feature has the same influence.

In [29]:
scaler = MinMaxScaler()
df_recommender_scaled = scaler.fit_transform(df_recommender)
df_recommender_scaled = pd.DataFrame(df_recommender_scaled, columns=df_recommender.columns, index = df_recommender.index)

In [30]:
df_recommender_scaled.head()

danceability    energy       key  loudness  mode  \
spotify_id                                                                 
000P83HDtOHcNVFZy7Q2Yu      0.638803  0.561519  0.636364  0.798337   0.0   
000paUq6xoYcvCz2REJdu1      0.507740  0.558516  0.181818  0.764918   1.0   
0017A9myIXhXYB4gBsPRsA      0.227038  0.668637  0.000000  0.901301   1.0   
001reycpvPkUjmpxTJv6eV      0.462332  0.224149  0.454545  0.662231   0.0   
002H0J6N0wKjjRtFQhKDzZ      0.295150  0.480430  0.181818  0.711602   0.0   

                        speechiness  acousticness  instrumentalness  liveness  \
spotify_id                                                                      
000P83HDtOHcNVFZy7Q2Yu     0.076507      0.439758          0.000042  0.114847   
000paUq6xoYcvCz2REJdu1     0.090229      0.743976          0.037851  0.212072   
0017A9myIXhXYB4gBsPRsA     0.030042      0.084336          0.000004  0.105732   
001reycpvPkUjmpxTJv6eV     0.036175      0.834337          0.919679  0.117885   
002H0J6N0wKjjRtFQhKDzZ     0.037942      0.530120          0.324297  0.334616   

                        valence     tempo  duration_ms  time_signature  
spotify_id                                                              
000P83HDtOHcNVFZy7Q2Yu    0.696  0.387169     0.067168             0.8  
000paUq6xoYcvCz2REJdu1    0.606  0.418638     0.104717             0.8  
0017A9myIXhXYB4gBsPRsA    0.360  0.354490     0.069198             0.8  
001reycpvPkUjmpxTJv6eV    0.540  0.488358     0.059419             0.8  
002H0J6N0wKjjRtFQhKDzZ    0.752  0.584514     0.097124             0.8

In [31]:
df_recommender_scaled.reset_index(inplace=True)
df_recommender.reset_index(inplace=True)

Convert every genre to a feature. If a song is part of a genre it should contain the value 1 otherwise 0.

In [32]:
ct = pd.crosstab(df['spotify_id'], df['genres'])
ct.reset_index(inplace=True)

In [33]:
display(ct.head())
ct.shape

genres              spotify_id  acoustic  afrobeat  alt-rock  alternative  \
0       000P83HDtOHcNVFZy7Q2Yu         0         0         0            0   
1       000paUq6xoYcvCz2REJdu1         0         0         0            0   
2       0017A9myIXhXYB4gBsPRsA         0         0         0            0   
3       001reycpvPkUjmpxTJv6eV         0         0         0            0   
4       002H0J6N0wKjjRtFQhKDzZ         0         0         0            0   

genres  ambient  anime  bluegrass  blues  brazil  ...  sleep  songwriter  \
0             0      0          0      0       0  ...      0           0   
1             0      0          0      0       0  ...      0           0   
2             0      0          0      0       0  ...      0           0   
3             0      0          0      0       0  ...      0           0   
4             0      0          0      0       0  ...      0           0   

genres  soul  spanish  swedish  synth-pop  tango  techno  trance  turkish  
0          0        0        0          0      0       0       0        0  
1          0        0        0          0      0       0       0        0  
2          0        0        0          0      0       0       0        0  
3          0        0        0          0      0       0       0        0  
4          0        0        0          0      0       0       0        1  

[5 rows x 96 columns]

(38456, 96)

Merge both DataFrames together to create the Recommender System.

In [34]:
ct_merged = ct.merge(df_recommender_scaled, on=['spotify_id'], how='left')

In [35]:
ct_merged.set_index(['spotify_id'], inplace=True)

In [36]:
display(ct_merged.head())
ct_merged.shape

acoustic  afrobeat  alt-rock  alternative  ambient  \
spotify_id                                                                   
000P83HDtOHcNVFZy7Q2Yu         0         0         0            0        0   
000paUq6xoYcvCz2REJdu1         0         0         0            0        0   
0017A9myIXhXYB4gBsPRsA         0         0         0            0        0   
001reycpvPkUjmpxTJv6eV         0         0         0            0        0   
002H0J6N0wKjjRtFQhKDzZ         0         0         0            0        0   

                        anime  bluegrass  blues  brazil  british  ...  \
spotify_id                                                        ...   
000P83HDtOHcNVFZy7Q2Yu      0          0      0       0        0  ...   
000paUq6xoYcvCz2REJdu1      0          0      0       0        0  ...   
0017A9myIXhXYB4gBsPRsA      0          0      0       0        0  ...   
001reycpvPkUjmpxTJv6eV      0          0      0       0        0  ...   
002H0J6N0wKjjRtFQhKDzZ      0          0      0       0        0  ...   

                        loudness  mode  speechiness  acousticness  \
spotify_id                                                          
000P83HDtOHcNVFZy7Q2Yu  0.798337   0.0     0.076507      0.439758   
000paUq6xoYcvCz2REJdu1  0.764918   1.0     0.090229      0.743976   
0017A9myIXhXYB4gBsPRsA  0.901301   1.0     0.030042      0.084336   
001reycpvPkUjmpxTJv6eV  0.662231   0.0     0.036175      0.834337   
002H0J6N0wKjjRtFQhKDzZ  0.711602   0.0     0.037942      0.530120   

                        instrumentalness  liveness  valence     tempo  \
spotify_id                                                              
000P83HDtOHcNVFZy7Q2Yu          0.000042  0.114847    0.696  0.387169   
000paUq6xoYcvCz2REJdu1          0.037851  0.212072    0.606  0.418638   
0017A9myIXhXYB4gBsPRsA          0.000004  0.105732    0.360  0.354490   
001reycpvPkUjmpxTJv6eV          0.919679  0.117885    0.540  0.488358   
002H0J6N0wKjjRtFQhKDzZ          0.324297  0.334616    0.752  0.584514   

                        duration_ms  time_signature  
spotify_id                                           
000P83HDtOHcNVFZy7Q2Yu     0.067168             0.8  
000paUq6xoYcvCz2REJdu1     0.104717             0.8  
0017A9myIXhXYB4gBsPRsA     0.069198             0.8  
001reycpvPkUjmpxTJv6eV     0.059419             0.8  
002H0J6N0wKjjRtFQhKDzZ     0.097124             0.8  

[5 rows x 108 columns]

(38456, 108)

Overall there are 108 features used for the Recommendation system.

Try different distance measures / similarity functions.

In [38]:
from sklearn.metrics.pairwise import cosine_similarity, linear_kernel, polynomial_kernel, sigmoid_kernel

Compute the cosine similarity between each song. A value close to one means that 2 two songs are equal a value close to 0 means, that they are completley different.

Cosine similarity

In [37]:
cosine_similarity_array = cosine_similarity(ct_merged)

In [39]:
cosine_similarity_array.shape

(38456, 38456)

In [40]:
cosine_similarity_df = pd.DataFrame(cosine_similarity_array, index=ct_merged.index, columns=ct_merged.index)

In [41]:
cosine_similarity_df

spotify_id              000P83HDtOHcNVFZy7Q2Yu  000paUq6xoYcvCz2REJdu1  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                1.000000                0.584950   
000paUq6xoYcvCz2REJdu1                0.584950                1.000000   
0017A9myIXhXYB4gBsPRsA                0.547657                0.850264   
001reycpvPkUjmpxTJv6eV                0.577832                0.483625   
002H0J6N0wKjjRtFQhKDzZ                0.615026                0.521155   
...                                        ...                     ...   
7zsH8mPq4nPBqqqDJpjgJe                0.656446                0.723518   
7zv3sIP0Umtz9MswoJSF1T                0.619765                0.538335   
7zwHfIlTAIPa6hgjzK1MGx                0.633239                0.676929   
7zwlOlX3MA7BFkmf0c1y4p                0.511864                0.567409   
7zzdiBe3usdBtyqHEekCEs                0.490575                0.531816   

spotify_id              0017A9myIXhXYB4gBsPRsA  001reycpvPkUjmpxTJv6eV  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                0.547657                0.577832   
000paUq6xoYcvCz2REJdu1                0.850264                0.483625   
0017A9myIXhXYB4gBsPRsA                1.000000                0.398664   
001reycpvPkUjmpxTJv6eV                0.398664                1.000000   
002H0J6N0wKjjRtFQhKDzZ                0.490335                0.554062   
...                                        ...                     ...   
7zsH8mPq4nPBqqqDJpjgJe                0.790129                0.491403   
7zv3sIP0Umtz9MswoJSF1T                0.437428                0.569164   
7zwHfIlTAIPa6hgjzK1MGx                0.627054                0.656092   
7zwlOlX3MA7BFkmf0c1y4p                0.495387                0.570760   
7zzdiBe3usdBtyqHEekCEs                0.591632                0.369105   

spotify_id              002H0J6N0wKjjRtFQhKDzZ  002Ra2MFY7R5ci6q1YLWLH  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                0.615026                0.733066   
000paUq6xoYcvCz2REJdu1                0.521155                0.757498   
0017A9myIXhXYB4gBsPRsA                0.490335                0.702231   
001reycpvPkUjmpxTJv6eV                0.554062                0.586211   
002H0J6N0wKjjRtFQhKDzZ                1.000000                0.576255   
...                                        ...                     ...   
7zsH8mPq4nPBqqqDJpjgJe                0.551722                0.853721   
7zv3sIP0Umtz9MswoJSF1T                0.543438                0.659313   
7zwHfIlTAIPa6hgjzK1MGx                0.599098                0.614871   
7zwlOlX3MA7BFkmf0c1y4p                0.459383                0.558395   
7zzdiBe3usdBtyqHEekCEs                0.575584                0.644105   

spotify_id              003787JNriexFRPKRwyMTn  003FCWl5ci9MxCaiTbErSr  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                0.436763                0.696065   
000paUq6xoYcvCz2REJdu1                0.552569                0.725537   
0017A9myIXhXYB4gBsPRsA                0.499048                0.689698   
001reycpvPkUjmpxTJv6eV                0.539577                0.549387   
002H0J6N0wKjjRtFQhKDzZ                0.427358                0.589067   
...                                        ...                     ...   
7zsH8mPq4nPBqqqDJpjgJe                0.366343                0.582482   
7zv3sIP0Umtz9MswoJSF1T                0.482791                0.584925   
7zwHfIlTAIPa6hgjzK1MGx                0.688482                0.765184   
7zwlOlX3MA7BFkmf0c1y4p                0.814217                0.625960   
7zzdiBe3usdBtyqHEekCEs                0.286651                0.419448   

spotify_id              004223m8Zb9ZHcxHFxlpQc  006XjVYo0HbGrUfvL4FodP  ...  \
spotify_id                                                           

Linear kernel

In [42]:
linear_kernel_array = linear_kernel(ct_merged)

In [43]:
linear_kernel_array.shape

(38456, 38456)

In [44]:
linear_kernel_df = pd.DataFrame(linear_kernel_array, index=ct_merged.index, columns=ct_merged.index)
linear_kernel_df.head()

spotify_id              000P83HDtOHcNVFZy7Q2Yu  000paUq6xoYcvCz2REJdu1  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                4.256930                2.953650   
000paUq6xoYcvCz2REJdu1                2.953650                5.989404   
0017A9myIXhXYB4gBsPRsA                2.324010                4.279828   
001reycpvPkUjmpxTJv6eV                2.831302                2.810841   
002H0J6N0wKjjRtFQhKDzZ                2.812805                2.827201   

spotify_id              0017A9myIXhXYB4gBsPRsA  001reycpvPkUjmpxTJv6eV  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                2.324010                2.831302   
000paUq6xoYcvCz2REJdu1                4.279828                2.810841   
0017A9myIXhXYB4gBsPRsA                4.230210                1.947260   
001reycpvPkUjmpxTJv6eV                1.947260                5.639912   
002H0J6N0wKjjRtFQhKDzZ                2.235485                2.916704   

spotify_id              002H0J6N0wKjjRtFQhKDzZ  002Ra2MFY7R5ci6q1YLWLH  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                2.812805                3.180754   
000paUq6xoYcvCz2REJdu1                2.827201                3.898633   
0017A9myIXhXYB4gBsPRsA                2.235485                3.037384   
001reycpvPkUjmpxTJv6eV                2.916704                2.927719   
002H0J6N0wKjjRtFQhKDzZ                4.913558                2.686284   

spotify_id              003787JNriexFRPKRwyMTn  003FCWl5ci9MxCaiTbErSr  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                2.155379                3.570523   
000paUq6xoYcvCz2REJdu1                3.234504                4.414538   
0017A9myIXhXYB4gBsPRsA                2.455009                3.526743   
001reycpvPkUjmpxTJv6eV                3.064924                3.243758   
002H0J6N0wKjjRtFQhKDzZ                2.265792                3.246359   

spotify_id              004223m8Zb9ZHcxHFxlpQc  006XjVYo0HbGrUfvL4FodP  ...  \
spotify_id                                                              ...   
000P83HDtOHcNVFZy7Q2Yu                3.197774                2.039307  ...   
000paUq6xoYcvCz2REJdu1                2.987895                3.091287  ...   
0017A9myIXhXYB4gBsPRsA                2.639685                2.276847  ...   
001reycpvPkUjmpxTJv6eV                2.612997                3.233974  ...   
002H0J6N0wKjjRtFQhKDzZ                2.864889                1.883516  ...   

spotify_id              7zn1gpif2DHi9t9isA5Boe  7znOzcWy38z63lIcKOQBd4  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                3.181161                2.756306   
000paUq6xoYcvCz2REJdu1                4.115933                3.525990   
0017A9myIXhXYB4gBsPRsA                3.557461                3.144690   
001reycpvPkUjmpxTJv6eV                2.963520                2.347248   
002H0J6N0wKjjRtFQhKDzZ                3.011848                2.388455   

spotify_id              7zndxriU3fd1TkV9BMdv5Y  7zraaKkjSygFzRAObStH8C  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                3.610660                3.685095   
000paUq6xoYcvCz2REJdu1                6.358596                4.395858   
0017A9myIXhXYB4gBsPRsA                4.401928                3.829674   
001reycpvPkUjmpxTJv6eV                3.288713                3.067704   
002H0J6N0wKjjRtFQhKDzZ                3.264198                3.273877   

spotify_id              7zrlCQpCrfAxWEop9J8OzK  7zsH8mPq4nPBqqqDJpjgJe  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                2.730652                2.445887   
000paUq6xoYcvCz2REJdu1                3.642021                3.197646   
0017A9myIXhXYB4gBsPRsA              

In [ ]:
df.sort_values('popularity', ascending=False).head(10)

In [51]:
def recommend_tracks(track: str, df_rec: pd.DataFrame):
    '''
    Recommends tracks that are similar to the provided track.

    Parameter
    ---------
    track: str
        Provided track

    df: pd.DataFrame
        DataFrame used for the Recommendation
    
    '''
    ids = df[df.name == track]['spotify_id'].values
    for id in ids:
        recommender_df = df_rec[[ids[0]]]
        recommender_df.reset_index(inplace=True)
        recommender_df = recommender_df.rename(columns={id: 'Score'})
        recommender_df = recommender_df.merge(df_total, how='left', on='spotify_id')
        display(recommender_df.sort_values(by='Score', ascending=False).head(10))


In [52]:
recommend_tracks('Back In Black', cosine_similarity_df)

spotify_id     Score             genres  \
722    08mG3Y1vljYA6bvDt4Wqkj  1.000000  [hard-rock, rock]   
25422  5AD7oNm0kvWIXxW3vy0z2a  0.996716  [hard-rock, rock]   
26519  5O4rVUjsG66l2nArhYW57m  0.996298  [hard-rock, rock]   
6199   1En7X8Ama1U8MosJbDndnB  0.992350  [hard-rock, rock]   
18529  3m3hmciMkd4y3XGC1mjTHf  0.991354  [hard-rock, rock]   
36341  7MxmJmW08fXXIx0WV1UeKz  0.990243  [hard-rock, rock]   
25670  5DVXOyNMl4MunWZ2BDvgA5  0.989644  [hard-rock, rock]   
25215  57hFsLrkALFFIxncLAY7Tg  0.989644  [hard-rock, rock]   
16911  3R306PKKOPAUjSSMkDHOpY  0.988208  [hard-rock, rock]   
21489  4NyZk1OfjwkGiIQxsD10ii  0.987438  [hard-rock, rock]   

                                                    name  \
722                       [Back In Black, Back In Black]   
25422                     [Homemade Love, Homemade Love]   
26519                           [Relentless, Relentless]   
6199                        [Too Far Gone, Too Far Gone]   
18529                 [Rolê da Estrada, Rolê da Estrada]   
36341                 [Someone Tonight, Someone Tonight]   
25670                     [Hello America, Hello America]   
25215                     [Hello America, Hello America]   
16911  [Rock n' Roll Is Alive and Well, Rock n' Roll ...   
21489                                 [Anytime, Anytime]   

                                        artists  \
722                              [AC/DC, AC/DC]   
25422                        [Journey, Journey]   
26519                          [Kansas, Kansas]   
6199                     [The Babys, The Babys]   
18529  [Patrulha do Espaço, Patrulha do Espaço]   
36341          [REO Speedwagon, REO Speedwagon]   
25670                [Def Leppard, Def Leppard]   
25215                [Def Leppard, Def Leppard]   
16911                    [The Babys, The Babys]   
21489                    [The Babys, The Babys]   

                                                   album  \
722                       [Back In Black, Back In Black]   
25422                             [Departure, Departure]   
26519                     [Audio-Visions, Audio-Visions]   
6199                          [On the Edge, On the Edge]   
18529           [Patrulha do Espaço, Patrulha do Espaço]   
36341  [Hi Infidelity (30th Anniversary Edition), Hi ...   
25670  [On Through The Night (Remastered), On Through...   
25215       [On Through The Night, On Through The Night]   
16911                         [On the Edge, On the Edge]   
21489                         [Union Jacks, Union Jacks]   

                   release_date release_date_precision  \
722    [1980-07-25, 1980-07-25]             [day, day]   
25422              [1980, 1980]           [year, year]   
26519  [1980-09-01, 1980-09-01]             [day, day]   
6199   [1980-10-01, 1980-10-01]             [day, day]   
18529  [1980-10-10, 1980-10-10]             [day, day]   
36341              [1980, 1980]           [year, year]   
25670  [1980-03-14, 1980-03-14]             [day, day]   
25215  [1980-03-14, 1980-03-14]             [day, day]   
16911  [1980-10-01, 1980-10-01]             [day, day]   
21489  [1980-01-01, 1980-01-01]             [day, day]   

                                                     uri  \
722    [spotify:track:08mG3Y1vljYA6bvDt4Wqkj, spotify...   
25422  [spotify:track:5AD7oNm0kvWIXxW3vy0z2a, spotify...   
26519  [spotify:track:5O4rVUjsG66l2nArhYW57m, spotify...   
6199   [spotify:track:1En7X8Ama1U8MosJbDndnB, spotify...   
18529  [spotify:track:3m3hmciMkd4y3XGC1mjTHf, spotify...   
36341  [spotify:track:7MxmJmW08fXXIx0WV1UeKz, spotify...   
25670  [spotify:track:5DVXOyNMl4MunWZ2BDvgA5, spotify...   
25215  [spotify:track:57hFsLrkALFFIxncLAY7Tg, spotify...   
16911  [spotify:track:3R306PKKOPAUjSSMkDHOpY, spotify...   
21489  [spotify:track:4NyZk1OfjwkGiIQxsD10ii, spotify...   

                               isrc  ...  loudness  mode  speechiness  \
722    [AUAP08000046, AUAP08000046]  ...    -5.678   1.0       0.0470   
25422 

spotify_id     Score             genres  \
722    08mG3Y1vljYA6bvDt4Wqkj  1.000000  [hard-rock, rock]   
25422  5AD7oNm0kvWIXxW3vy0z2a  0.996716  [hard-rock, rock]   
26519  5O4rVUjsG66l2nArhYW57m  0.996298  [hard-rock, rock]   
6199   1En7X8Ama1U8MosJbDndnB  0.992350  [hard-rock, rock]   
18529  3m3hmciMkd4y3XGC1mjTHf  0.991354  [hard-rock, rock]   
36341  7MxmJmW08fXXIx0WV1UeKz  0.990243  [hard-rock, rock]   
25670  5DVXOyNMl4MunWZ2BDvgA5  0.989644  [hard-rock, rock]   
25215  57hFsLrkALFFIxncLAY7Tg  0.989644  [hard-rock, rock]   
16911  3R306PKKOPAUjSSMkDHOpY  0.988208  [hard-rock, rock]   
21489  4NyZk1OfjwkGiIQxsD10ii  0.987438  [hard-rock, rock]   

                                                    name  \
722                       [Back In Black, Back In Black]   
25422                     [Homemade Love, Homemade Love]   
26519                           [Relentless, Relentless]   
6199                        [Too Far Gone, Too Far Gone]   
18529                 [Rolê da Estrada, Rolê da Estrada]   
36341                 [Someone Tonight, Someone Tonight]   
25670                     [Hello America, Hello America]   
25215                     [Hello America, Hello America]   
16911  [Rock n' Roll Is Alive and Well, Rock n' Roll ...   
21489                                 [Anytime, Anytime]   

                                        artists  \
722                              [AC/DC, AC/DC]   
25422                        [Journey, Journey]   
26519                          [Kansas, Kansas]   
6199                     [The Babys, The Babys]   
18529  [Patrulha do Espaço, Patrulha do Espaço]   
36341          [REO Speedwagon, REO Speedwagon]   
25670                [Def Leppard, Def Leppard]   
25215                [Def Leppard, Def Leppard]   
16911                    [The Babys, The Babys]   
21489                    [The Babys, The Babys]   

                                                   album  \
722                       [Back In Black, Back In Black]   
25422                             [Departure, Departure]   
26519                     [Audio-Visions, Audio-Visions]   
6199                          [On the Edge, On the Edge]   
18529           [Patrulha do Espaço, Patrulha do Espaço]   
36341  [Hi Infidelity (30th Anniversary Edition), Hi ...   
25670  [On Through The Night (Remastered), On Through...   
25215       [On Through The Night, On Through The Night]   
16911                         [On the Edge, On the Edge]   
21489                         [Union Jacks, Union Jacks]   

                   release_date release_date_precision  \
722    [1980-07-25, 1980-07-25]             [day, day]   
25422              [1980, 1980]           [year, year]   
26519  [1980-09-01, 1980-09-01]             [day, day]   
6199   [1980-10-01, 1980-10-01]             [day, day]   
18529  [1980-10-10, 1980-10-10]             [day, day]   
36341              [1980, 1980]           [year, year]   
25670  [1980-03-14, 1980-03-14]             [day, day]   
25215  [1980-03-14, 1980-03-14]             [day, day]   
16911  [1980-10-01, 1980-10-01]             [day, day]   
21489  [1980-01-01, 1980-01-01]             [day, day]   

                                                     uri  \
722    [spotify:track:08mG3Y1vljYA6bvDt4Wqkj, spotify...   
25422  [spotify:track:5AD7oNm0kvWIXxW3vy0z2a, spotify...   
26519  [spotify:track:5O4rVUjsG66l2nArhYW57m, spotify...   
6199   [spotify:track:1En7X8Ama1U8MosJbDndnB, spotify...   
18529  [spotify:track:3m3hmciMkd4y3XGC1mjTHf, spotify...   
36341  [spotify:track:7MxmJmW08fXXIx0WV1UeKz, spotify...   
25670  [spotify:track:5DVXOyNMl4MunWZ2BDvgA5, spotify...   
25215  [spotify:track:57hFsLrkALFFIxncLAY7Tg, spotify...   
16911  [spotify:track:3R306PKKOPAUjSSMkDHOpY, spotify...   
21489  [spotify:track:4NyZk1OfjwkGiIQxsD10ii, spotify...   

                               isrc  ...  loudness  mode  speechiness  \
722    [AUAP08000046, AUAP08000046]  ...    -5.678   1.0       0.0470   
25422 

In [53]:
recommend_tracks('Back In Black', linear_kernel_df)

spotify_id     Score  \
2112   0PLzoe4izqluPTVsF9K4mb  7.155934   
4650   0vEEhp1BFWt1iCKsojoebd  7.150820   
6199   1En7X8Ama1U8MosJbDndnB  7.114456   
1578   0Ix7jl28ovNUGclUmkwHnx  7.092672   
28209  5joQhLd0l0jc1S0U9TkmsG  7.073923   
29747  63YhDDyYQJo847NqN7mry3  7.066416   
17610  3Zw7u9z5sfbasLy9zkubzn  7.045456   
30396  6BUuoR1kG1rRag718ibuOn  7.043061   
12883  2cXADTZcYNhvtfZuK77Xt0  7.040993   
1883   0MLxcHeTU7mJWAKbbwOPT5  7.034511   

                                                  genres  \
2112        [hard-rock, hardcore, punk, punk-rock, rock]   
4650        [hard-rock, hardcore, punk, punk-rock, rock]   
6199                                   [hard-rock, rock]   
1578                                   [hard-rock, rock]   
28209  [alt-rock, alternative, blues, hard-rock, hard...   
29747  [alt-rock, alternative, blues, grunge, hard-ro...   
17610           [blues, british, hard-rock, metal, rock]   
30396                   [country, folk, hard-rock, rock]   
12883                    [blues, hard-rock, metal, rock]   
1883   [alt-rock, alternative, hard-rock, hardcore, p...   

                                                    name  \
2112   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
4650                 [Ska 4, Ska 4, Ska 4, Ska 4, Ska 4]   
6199                        [Too Far Gone, Too Far Gone]   
1578                                  [My Life, My Life]   
28209  [Soul Kitchen, Soul Kitchen, Soul Kitchen, Sou...   
29747  [Cover Band - Demo, Cover Band - Demo, Cover B...   
17610  [Carry Your Load, Carry Your Load, Carry Your ...   
30396  [Betty Lou's Gettin' Out Tonight, Betty Lou's ...   
12883  [Hungry Boys, Hungry Boys, Hungry Boys, Hungry...   
1883   [Let's Go Die, Let's Go Die, Let's Go Die, Let...   

                                                 artists  \
2112   [Maukka Perusjätkä,Ralf Örn Pop, Maukka Perusj...   
4650   [Maukka Perusjätkä,Ralf Örn Pop, Maukka Perusj...   
6199                              [The Babys, The Babys]   
1578                          [Peter Criss, Peter Criss]   
28209                        [X, X, X, X, X, X, X, X, X]   
29747  [Redd Kross, Redd Kross, Redd Kross, Redd Kros...   
17610  [Whitesnake, Whitesnake, Whitesnake, Whitesnak...   
30396       [Bob Seger, Bob Seger, Bob Seger, Bob Seger]   
12883  [Blue Öyster Cult, Blue Öyster Cult, Blue Öyst...   
1883   [Hüsker Dü, Hüsker Dü, Hüsker Dü, Hüsker Dü, H...   

                                                   album  \
2112   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
4650   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
6199                          [On the Edge, On the Edge]   
1578                    [Out Of Control, Out Of Control]   
28209  [Los Angeles, Los Angeles, Los Angeles, Los An...   
29747  [Red Cross, Red Cross, Red Cross, Red Cross, R...   
17610  [Ready An' Willing, Ready An' Willing, Ready A...   
30396  [Against The Wind, Against The Wind, Against T...   
12883  [Cultosaurus Erectus, Cultosaurus Erectus, Cul...   
1883   [Statues / Amusement, Statues / Amusement, Sta...   

                                            release_date  \
2112   [1980-12-12, 1980-12-12, 1980-12-12, 1980-12-1...   
4650   [1980-12-12, 1980-12-12, 1980-12-12, 1980-12-1...   
6199                            [1980-10-01, 1980-10-01]   
1578                                        [1980, 1980]   
28209  [1980-04-26, 1980-04-26, 1980-04-26, 1980-04-2...   
29747  [1980-01-01, 1980-01-01, 1980-01-01, 1980-01-0...   
17610  [1980-01-01, 1980-01-01, 1980-01-01, 1980-01-0...   
30396   [1980-02-25, 1980-02-25, 1980-02-25, 1980-02-25]   
12883                           [1980, 1980, 1980, 1980]   
1883   [1980-08-01, 1980-08-01, 1980-08-01, 1980-08-0...   

                                  release_date_precision  \
2112                           [day, day, day, day, day]   
4650                           [day, day, day, day, day]   
6199                                          [day, day]   


spotify_id     Score  \
2112   0PLzoe4izqluPTVsF9K4mb  7.155934   
4650   0vEEhp1BFWt1iCKsojoebd  7.150820   
6199   1En7X8Ama1U8MosJbDndnB  7.114456   
1578   0Ix7jl28ovNUGclUmkwHnx  7.092672   
28209  5joQhLd0l0jc1S0U9TkmsG  7.073923   
29747  63YhDDyYQJo847NqN7mry3  7.066416   
17610  3Zw7u9z5sfbasLy9zkubzn  7.045456   
30396  6BUuoR1kG1rRag718ibuOn  7.043061   
12883  2cXADTZcYNhvtfZuK77Xt0  7.040993   
1883   0MLxcHeTU7mJWAKbbwOPT5  7.034511   

                                                  genres  \
2112        [hard-rock, hardcore, punk, punk-rock, rock]   
4650        [hard-rock, hardcore, punk, punk-rock, rock]   
6199                                   [hard-rock, rock]   
1578                                   [hard-rock, rock]   
28209  [alt-rock, alternative, blues, hard-rock, hard...   
29747  [alt-rock, alternative, blues, grunge, hard-ro...   
17610           [blues, british, hard-rock, metal, rock]   
30396                   [country, folk, hard-rock, rock]   
12883                    [blues, hard-rock, metal, rock]   
1883   [alt-rock, alternative, hard-rock, hardcore, p...   

                                                    name  \
2112   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
4650                 [Ska 4, Ska 4, Ska 4, Ska 4, Ska 4]   
6199                        [Too Far Gone, Too Far Gone]   
1578                                  [My Life, My Life]   
28209  [Soul Kitchen, Soul Kitchen, Soul Kitchen, Sou...   
29747  [Cover Band - Demo, Cover Band - Demo, Cover B...   
17610  [Carry Your Load, Carry Your Load, Carry Your ...   
30396  [Betty Lou's Gettin' Out Tonight, Betty Lou's ...   
12883  [Hungry Boys, Hungry Boys, Hungry Boys, Hungry...   
1883   [Let's Go Die, Let's Go Die, Let's Go Die, Let...   

                                                 artists  \
2112   [Maukka Perusjätkä,Ralf Örn Pop, Maukka Perusj...   
4650   [Maukka Perusjätkä,Ralf Örn Pop, Maukka Perusj...   
6199                              [The Babys, The Babys]   
1578                          [Peter Criss, Peter Criss]   
28209                        [X, X, X, X, X, X, X, X, X]   
29747  [Redd Kross, Redd Kross, Redd Kross, Redd Kros...   
17610  [Whitesnake, Whitesnake, Whitesnake, Whitesnak...   
30396       [Bob Seger, Bob Seger, Bob Seger, Bob Seger]   
12883  [Blue Öyster Cult, Blue Öyster Cult, Blue Öyst...   
1883   [Hüsker Dü, Hüsker Dü, Hüsker Dü, Hüsker Dü, H...   

                                                   album  \
2112   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
4650   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
6199                          [On the Edge, On the Edge]   
1578                    [Out Of Control, Out Of Control]   
28209  [Los Angeles, Los Angeles, Los Angeles, Los An...   
29747  [Red Cross, Red Cross, Red Cross, Red Cross, R...   
17610  [Ready An' Willing, Ready An' Willing, Ready A...   
30396  [Against The Wind, Against The Wind, Against T...   
12883  [Cultosaurus Erectus, Cultosaurus Erectus, Cul...   
1883   [Statues / Amusement, Statues / Amusement, Sta...   

                                            release_date  \
2112   [1980-12-12, 1980-12-12, 1980-12-12, 1980-12-1...   
4650   [1980-12-12, 1980-12-12, 1980-12-12, 1980-12-1...   
6199                            [1980-10-01, 1980-10-01]   
1578                                        [1980, 1980]   
28209  [1980-04-26, 1980-04-26, 1980-04-26, 1980-04-2...   
29747  [1980-01-01, 1980-01-01, 1980-01-01, 1980-01-0...   
17610  [1980-01-01, 1980-01-01, 1980-01-01, 1980-01-0...   
30396   [1980-02-25, 1980-02-25, 1980-02-25, 1980-02-25]   
12883                           [1980, 1980, 1980, 1980]   
1883   [1980-08-01, 1980-08-01, 1980-08-01, 1980-08-0...   

                                  release_date_precision  \
2112                           [day, day, day, day, day]   
4650                           [day, day, day, day, day]   
6199                                          [day, day]   


In [54]:
import pandas as pd

In [55]:
colums = ['Vorname', 'Nachname']

df = pd.DataFrame([['Jannis', 'Weichler'], ['Lite', 'Brandt']], columns=colums)

In [61]:
df.to_csv('names.csv', index=False)
df.to_csv('names.csv', index=False)

In [62]:
pd.read_csv('names.csv')

Vorname  Nachname
0  Jannis  Weichler
1    Lite    Brandt